api/src/main/java/bio/terra/tanagra/service/search/Filter.java

```diff
@@ -74,5 +76,66 @@ public static ArrayFunction create(List<Filter> operands, Operator operator) {
     }
   }
 
-  // TODO implement a relationship filter tht allows chaining to another entity.
+  /**
+   * A {@link Filter} that introduces a constraint that there must be a relationship to another
+   * entity.
+   *
+   * <p>This quantifiers introduces a new {@link Variable} and implicitly binds it to an entity in
+   * {@link #boundAttribute()}.
+   *
+   * <p>This is equivalent to the pseudo-SQL ":outerAttribute IN (SELECT :boundAttribute FROM
+   * :boundAttribute.entity AS :boundAttribute.variable WHERE :filter".
+   *
+   * <p>Or in words ∃x∈boundEntity( x.boundAttribute = outerAttribute ∧ filter): there exists {@link
+   * #boundAttribute()} that's a member of {@link #boundAttribute()}'s Entity where {@link
+   * #boundAttribute()} equals {@link #outerAttribute()} and {@link #filter}.
+   */
+  @AutoValue
+  abstract class RelationshipFilter implements Filter {
+
+    /**
+     * The attribute of the newly introduced variable's entity that is equal to {@link
+     * #outerAttribute}.
+     */
+    public abstract AttributeVariable boundAttribute();
+
+    /** The attribute of an outer variable that's equal to {@link #boundAttribute}. */
```

Comment:
Are there any restrictions on what entities are valid for this attribute? For example, if we have a heavily nested structure, can this attribute be defined multiple levels above this relationship filter, or just in the immediate surrounding context?

Is there any validation (implicit or explicit) here if I provide a reference to an entity which would not be valid in this position within the tree?

Things are getting a bit abstract so my terminology may be slightly off here.